[영화 극장 선택] 하디 1, 2단계 미션 제출합니다.

ii2001 opened the pull request:
안녕하세요 페로로! 또 만나네요

안드로이드 개발이 서툴러서 페어들이 잘 알려주었어요..!
잘 모르기때문에 저번처럼 차근차근 알려주시면 매우 감사하겠습니다..!
이번 미션에서도 페로로를 통해서 실력향상이 많이 되었으면 좋겠습니다
화이팅 해보겠습니다! 페로로도 화이팅!

🚀: 미션 수행 완료
👀: 고민중
❤️: 개꿀..?!


https://github.com/woowacourse/android-movie-theater/assets/74256335/fab2c062-045c-4d21-b18b-c3a77ce48ad0

woowahan-pjs committed:
:tada: Theater시작!

Commit:
:sparkles: 영화 선택 후 극장을 선택할 수 있음,
극장 이름 줄이기

Commit:
:recycle: 분기처리

Commit:
:recycle: dataBinding 적용

Commit:
:truck: seat->activity_theater_seat

Commit:
:truck: seat->activity_theater_seat

Commit:
:recycle: Test오류 고침

Commit:
:sparkles: 영화 시간 정보 최종 화면에 표시

murjune committed:
build: resources 파일 exclude 할 목록 추가

murjune committed:
ui: 바텀 네비 menu

murjune committed:
ui: 예약, 영화 목록, 세팅 Fragment

murjune committed:
ui: Fragment transaction 적용

murjune committed:
feat: MovieListFragment 로 마이그레이션

Commit:
:bug: 카운팅 안되던 버그 고침

murjune committed:
:recycle:  findFragmentByTag 로 개선

murjune committed:
:bug: ticket count 로직 개선

murjune committed:
:memo: 기능 요구사항 명세서 업데이트

murjune committed:
:white_check_mark: MovieListFragment test

murjune committed:
:truck: 이름 변경

murjune committed:
:truck: xml 이름 통일성

murjune committed:
:sparkles: timeSpinner 개선

murjune committed:
:recycle: TheaterViewHolder dataBinding 적용

murjune committed:
:recycle: MovieDetailActivity dataBinding으로 개선

murjune committed:
:recyle: seat activity databinding apply

murjune committed:
:recyle: Detail layout Spinner 위치 조정

murjune committed:
:recycle: :white_check_mark: PurchaseConfirmationActivityTest 구조 변경

- 액티비티 변경에 따른 테스트 코드 변경

murjune committed:
:recycle: :white_check_mark: MovieListFragmentTest 구조 변경

murjune committed:
:recycle: :white_check_box: 테스트 코드 고침

murjune committed:
:recycle: ktLint Check

laco-dev commented on README.md:
```diff
@@ -1,49 +1,21 @@
-- [x] 주말 에는 오전 9시부터 두 시간 간격으로 상영 한다.
-- [x] 평일 에는 오전 10시부터 두 시간 간격으로 상영 한다.
-- [x] 날짜와 시간은 기본 값으로 초기화 되어있다.
+- [ ]: dataBinding 적용
+- [ ]: test 코드 작성
+- [ ]: 화면 회전 대응
```
구현이 된 내용일까요? 아니면?

ii2001 replied:
구현이 되었는데 표시를 깜빡한것 같습니다 죄송합니다!

laco-dev commented on app/src/androidTest/java/woowacourse/movie/MovieDetailActivityTest.kt:
```diff
@@ -11,51 +13,75 @@ import androidx.test.ext.junit.runners.AndroidJUnit4
-    fun `초기값은1_마이너스버튼클릭_티켓수량0표시`() {
-        Espresso.onView(withId(R.id.minus_button))
-            .check(matches(isDisplayed()))
+    private val cinema
```
woowacourse.movie 테스트 코드의 패키지도 프로덕트와 동일하게 맞춰주세요.
이하 모두 동일합니다.

laco-dev commented on app/src/androidTest/java/woowacourse/movie/MovieListFragmentTest.kt:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Test
+    @DisplayName("영화 버튼 누르면 Dialog 보이는지 테스트")
+    fun test2() {
+        launchFragmentInContainer<MovieListFragment>()
+        onView(withId(R.id.rv_movies))
+            .perform(
+                RecyclerViewActions
+                    .actionOnItemAtPosition<MovieAdapter.MovieViewHolder>(
+                        0,
+                        clickItemWithId(R.id.movie_details_button),
+                    ),
+            )
+
+        onView(withId(R.id.rv_cinema)).check(matches(isDisplayed()))
+    }
```
다이얼로그가 극장 선택하는 Dialog를 말하는 것일까요?

영화 목록 프래그먼트의 테스트 범위로 봐야 하는가? 라는 의문이 들어서요

ii2001 replied:
통합 테스트라는 테스트파일을 만들어서 패키지 이동했습니다

laco-dev replied:
통합 테스트라는 파일 --> 통합을 할 정도로 규모가 있는 수준의 프로젝트가 아니에요.
각 테스트에 맞는 코드로 파일을 나눠주세요

laco-dev commented on app/src/main/java/woowacourse/movie/common/ui/ActivityExtensions.kt:
```diff
@@ -0,0 +1,21 @@
+import androidx.fragment.app.commit
+import androidx.fragment.app.replace
+
+inline fun <reified T : Fragment> FragmentActivity.replaceTo(
+    @IdRes fragmentContainerId: Int,
+    tag: String? = T::class.java.canonicalName,
+    args: Bundle? = null,
+    action: () -> Unit = {},
+) {
+    supportFragmentManager.commit {
+        replace<T>(fragmentContainerId, tag, args)
+        action()
+        setReorderingAllowed(true)
+    }
+}
```
이 함수는 정말 common이 맞을까요? 다른 화면에서도 충분히 재사용 될 수 있을정도로 의미있는 분리가 되었나요?

ii2001 replied:
제가 생각했을 때는 안드로이드 프레임워크(compose X) 를 사용하는 다른 화면이나 feature 에서도 Fragment 간 이동할 때 Transaction을  사용할 것 같습니다.
따라서, 공통적으로 해당 함수를 사용해 common package내에 존재하는게 맞다고 생각했습니다.
혹시 common이 아니라 util이라는 네임으로 변경한다면 느끼기에 괜찮을까요?
페로로 생각도 궁금하네요..!

laco-dev replied:
util도 마찬가지입니다.
재사용이 발생하는 것에는 동의 합니다. 하지만 그것을 무조건 util이나 common 처럼 공통 이라는 것으로 먼저 생각하는 것이 잘못되었습니다.

우리는 객체를 만들 때 역할을 정의하고, 그 역할을 구현할 클래스를 만듭니다.
그렇다면 프래그먼트 매니저는 왜 유틸이 아닐까요? 여러군데서 사용하는데도 불구하구요.

하고싶은 말은 공통 코드를 만든다는 개념으로 접근하는 것이 아닌, 트랜잭션을 도와주는 역할을 만든다고 해주시면 좋을 것 같아요.

laco-dev commented on app/src/main/java/woowacourse/movie/common/ui/FragmentExtensions.kt:
```diff
@@ -0,0 +1,10 @@
+import android.os.Bundle
+import androidx.fragment.app.Fragment
+
+inline fun <T : Fragment> T.withArgs(argsBuilder: Bundle.() -> Unit): T {
+    return this.apply {
+        arguments = Bundle().apply(argsBuilder)
+    }
+}
```
이 함수는 나름... common 처럼 느껴지네요!
다만 다른 개발자들이 이 함수가 있다는 사실 자체를 모를 것 같아서 아쉽게 느껴집니다.

Activity를 실행하는 함수를 정적 함수로 만드는 것처럼, 프래그먼트도 비슷한 방법이 가능합니다.

```kotlin
// Fragment

companion object {
   fun show(containerViewId: Int, args: Pair<String, Any>) { arguments = bundleOf(...) }
}
```

ii2001 replied:
위와 같습니다

ii2001 replied:
그러면 혹시 제가 우테코가 안드로이드 처음이라 3주차라 그런데
개발자들이 저렇게 common이라는 패키지 안에다가 함수를 만들어 놓으면 보지 않나요??
아니면 companion object 만드는게 오히려 더 확인이 쉬울까요?

laco-dev replied:
common 패키지에 파일이 200개 이렇게 있으면 매번 찾아서 볼 수 있을까요?
앞의 의견과 마찬가지로 무조건 유틸부터 만드는 것은 좋지 않은 습관이라서 그래요.

실제로 지금 이 코드는 1번만 사용되는걸요.

laco-dev commented on app/src/main/java/woowacourse/movie/model/Cinema.kt:
```diff
@@ -0,0 +1,9 @@
+import woowacourse.movie.model.theater.Theater
+import java.io.Serializable
+
+data class Cinema(
+    val cinemaName: String,
+    val theater: Theater,
+) : Serializable
```
Cinema와 Threater는 어떤 개념적 차이가 있나요?
제가 느끼기에는 극장이 더 크게 느껴져서요.

ii2001 replied:
Cinema: CGV, 롯데시네마 등 극장 이름이고
Theater: 상영관 입니다! 현재 1관, 2관 등 상영관을 정하는 요구사항이 없어서 우선 이렇게 네이밍 해놨습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/model/theater/Theater.kt:
```diff
@@ -2,5 +2,10 @@ package woowacourse.movie.model.theater
+import java.time.LocalTime
 
-class Theater(val movie: MovieInfo, private val seats: Map<String, Seat>) : Serializable
+data class Theater(
+    val movie: MovieInfo,
+    val times: List<LocalTime>,
+    private val seats: Map<String, Seat>,
+) : Serializable
```
하나의 극장이 하나의 영화정보와 시간들, 좌석들을 가지는 구조가 맞을까요?
강의 자료의 예시는 이렇게 나와있는데요.
```
선릉 극장
  스타워즈(09시, 11시, 15시)
  해리포터(13시, 15시, 17시, 19시)
  
잠실 극장
  스즈메의 문단속(09시)
  해리포터(10시, 12시, 14시)
  범죄도시(11시, 13시)
```

마찬가지로 극장 하나가 여러개의 상영 정보를 가지는게 현실세계에 비유할 때에도 더 어울리지 않을까요?

ii2001 replied:
현재 상영관 하나 당 영화 정보가 하나씩만 들어가게 되어있는 것 같습니다!
딱히 1,2, 4DX관등 선택하는게 없어서 이렇게 해놨습니다.
후에 여러개의 관이 생긴다면 movie와 times를 묶어 List로 묶을 생각입니다!
좌석리스트는 Theater에 있는 이유가 관마다 좌석수는 고정이라서 이렇게 해놨습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailActivity.kt:
```diff
@@ -1,69 +1,54 @@
+        setContentView(binding.root)
         supportActionBar?.setDisplayHomeAsUpEnabled(true)
+        val cinema =
+            IntentCompat.getSerializableExtra(intent, "Cinema", Cinema::class.java) ?: error(" ")
```
앱개발자가 가장 조심해야 할 것 중 하나는, 앱 사용중에 절대 크래시가 발생하지 않도록 하는 것인데요.
논리적인 오류여도 비정상 종료가 발생하지 않도록 만들어보세요

laco-dev commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailActivity.kt:
```diff
@@ -1,69 +1,54 @@
-    private val minusButton: Button by lazy { findViewById(R.id.minus_button) }
-    private val seatConfirmationButton: Button by lazy { findViewById(R.id.seat_confirmation_button) }
-    private val quantityText: TextView by lazy { findViewById(R.id.quantity_text_view) }
+    private lateinit var times: List<LocalTime>
```
이 데이터는 프레젠터가 가지고 있어야 할 내용이 아닐까요?

laco-dev commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailActivity.kt:
```diff
@@ -1,69 +1,54 @@
-            .setImageDrawable(ContextCompat.getDrawable(this, R.drawable.movie_making_poster))
+            ).also { binding.presenter = it }
+        presenter.load(theater.movie)
+        cinema.let { setupEventListeners(it) }
```
let을 사용하는 올바른 방법이 아닙니다.
단순히 변수를 편하게 사용하려고 하는 것은 잘못된 방법이에요.

범위 지정 함수들의 예시를 찾아보고 적절한 곳에 사용될 수 있도록 해주세요

laco-dev commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailContract.kt:
```diff
@@ -21,21 +17,18 @@ interface MovieDetailContract {
+        fun getTickets(): Int
 
-        fun getTicketNum(): Int
+        fun getMovie(): MovieInfo
```
MVP 패턴의 기본 원칙은, 프레젠터는 무언가 값을 반환하지 않아야 합니다.

뷰와 프레젠터는 서로간의 단방향으로 행동을 요청하기 때문에, 값을 꺼내오는 동작을 하는 대신 필요한 행동을 다시 인터페이스로 호출하도록 해야 합니다.

```
View -> UI Event -> Presenter -> Model -> Presenter -> View -> ...
```

laco-dev commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailContract.kt:
```diff
@@ -21,21 +17,18 @@ interface MovieDetailContract {
+        fun onTicketPlusClicked()
 
-        fun getTheater(): Theater?
+        fun onTicketMinusClicked()
```
또한 프레젠터는 UI 종속적으로 작성하지 않습니다.
클릭과 같은 동작은 UI에서만 사용되는 개념이고 프레젠터는 모델과 통신하는 주체이기 때문에 기능에 맞는 이름으로 작성되어야만 합니다.

가령 티켓 플러스 버튼 클릭이 아닌 더블 클릭 스와이프 터치, 드래그 이런식으로 바뀌더라도 인터페이스가 변경되지 않아야 합니다.

laco-dev commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailPresenter.kt:
```diff
@@ -1,70 +1,49 @@
+    override fun getMovie(): MovieInfo = movie
+
+    override fun load(movie: MovieInfo) {
+        this.movie = movie
```
영화를 불러오기 위해 영화를 파라미터로 넣는게 정상일까요??

laco-dev commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailPresenter.kt:
```diff
@@ -1,70 +1,49 @@
-        endDate: LocalDate,
-    ) {
+    override fun generateDateRange() {
+        val startDate = LocalDate.of(2024, 4, 1)
+        val endDate = LocalDate.of(2024, 4, 30)
```
어떤 목적을 위한 하드코딩일지는 모르겠어요.
다만 프레젠터에 직접 하드코딩을 남기는 대신 모델을 통해 이 값이 결정되어야 할 것 같습니다.

ii2001 replied:
제 생각엔 본래라면 서버에서 받아와야 하는 데이터라고 생각해서 그 부분을 그저 하드코딩 해놓은 겁니다..
모델로 초기화를 하기보단 데이터를 받아온다고 생각하면 괜찮지 않을까요?

laco-dev replied:
> 제 생각엔 본래라면 서버에서 받아와야 하는 데이터라고 생각해서 그 부분을 그저 하드코딩 해놓은 겁니다..

라는 것을 제가 가지고 있는 정보로 파악할 수 있을까요?
개발자는 기본적으로 코드로 대화를 합니다.

최소한 함수 혹은 클래스로 분리하는 노력을 할 수는 있습니다. 🙇🏻 
하다못해 주석이라도 남길 수 있어야 다른 개발자가 의도를 이해할 수 있습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/movieList/MovieListView.kt:
```diff
@@ -6,5 +6,5 @@ interface MovieListView {
     fun updateAdapter(displayData: List<MovieDisplayData>)
 
-    fun navigateToMovieDetail(theater: Theater)
+    fun showBottomSheet(theater: Theater)
```
showBottomSheet() 라는 동작 또한 UI가 구체적으로 어떻게 행동할지와 너무 자세하게 의존됩니다.
가령 바텀시트를 띄울 수도, 화면을 이동할 수도 토스트를 띄울지도 모릅니다.

laco-dev commented on app/src/main/java/woowacourse/movie/movieList/MovieListFragment.kt:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    private fun initPresenter() {
+        _presenter = MovieListPresenter(this)
+        presenter.loadMovies()
```
loadMovies() 동작은 프레젠터 생성자 블록에서 할 수 있지 않을까요?

ii2001 replied:
여러가지 방법으로 로드 해보겠습니다

laco-dev commented on app/src/main/java/woowacourse/movie/movieList/cinemaListDialog/CinemaAdapter.kt:
```diff
@@ -0,0 +1,49 @@
+
+class CinemaAdapter(private val onTheaterClicked: (Cinema) -> Unit) :
+    ListAdapter<Cinema, TheaterViewHolder>(theaterComparator) {
+    private lateinit var inflater: LayoutInflater
```
inflater를 어댑터의 클래스 필드로 가지고 있으면 어떤 장점이 있나요?
Inflater를 만드는 동작보다 실제로 inflate 하는 동작이 더 무겁습니다.

ii2001 replied:
onCreateViewHolder()가 호출될때마다 inflater를 만드는것 보다 이게 효율적이라고 생각했습니다

laco-dev replied:
스스로 효율적일 것이라고 판단한 점은 좋습니다.
다만 lateinit으로 남아있는 만큼 이 필드가 만들어 낼 수 있는 잠재적인 오류 발생 가능성은 무시할 수 없습니다.

싱글톤 객체 혹은 로또 번호를 만드는 것 처럼, 함수를 통해 가져오도록 할 수도 있습니다.
```kotlin
private var cachedInflate: Inflater?

fun getInflater(context: Context) { 
   cachedInflater ?: context.getInflater(...)
}
```
이 함수를 사용하는 입장에서는 더이상 lateinit으로 인해 문제가 발생할 것을 걱정하지 않아도 됩니다.

laco-dev commented on app/src/main/java/woowacourse/movie/movieList/cinemaListDialog/TheatersBottomSheetFragment.kt:
```diff
@@ -0,0 +1,98 @@
+        return binding.root
+    }
+
+    override fun onViewCreated(
+        view: View,
+        savedInstanceState: Bundle?,
+    ) {
```
onViewCreated() 함수가 다소 복잡합니다. 다이어트 해주세요 💪🏻

laco-dev commented on app/src/main/java/woowacourse/movie/movieList/cinemaListDialog/TheatersBottomSheetFragment.kt:
```diff
@@ -0,0 +1,98 @@
+            }
+        binding.rvCinema.adapter = adapter
+        adapter.submitList(
+            listOf(
+                Cinema(
+                    "CGV",
+                    theater.copy(
+                        times = listOf(LocalTime.of(9, 0), LocalTime.of(11, 0), LocalTime.of(15, 0)),
+                    ),
+                ),
```
데이터를 준비하는 것은, 프레젠터와 모델의 역할이 아닐까요? 하드코딩을 하는 것이 미션이 의도했던 요구사항일까요?

ii2001 replied:
Presenter로 이동했습니다

laco-dev commented on app/src/main/java/woowacourse/movie/purchaseConfirmation/PurchaseConfirmationActivity.kt:
```diff
@@ -2,25 +2,32 @@ package woowacourse.movie.purchaseConfirmation
+import woowacourse.movie.model.Cinema
 
 class PurchaseConfirmationActivity : AppCompatActivity() {
+    private val binding: ActivityPurchaseConfirmationBinding by lazy {
+        ActivityPurchaseConfirmationBinding.inflate(layoutInflater)
+    }
```
이 화면은 왜 MVP 구조로 안만들어졌을까요?

ii2001 replied:
프레젠터가 해야하는 일이 마땅히 없다고 생각해서 이런 구조로 만들었습니다.

laco-dev replied:
> "마땅히 없어서 안했다. 내 생각에는 안해도 될 것 같다"

개인 프로젝트에서는 이해할 수 있지만, 팀 단위 혹은 아키텍처를 지키는 구조에서는 조금 아쉬운 답변입니다.
내 마음대로 해도 된다면, 아키텍처가 있을 필요가 없으니까요?

혹은 이 화면이 이후 비즈니스 로직이 생겨나게 되면, 이렇게 생각할 지도 모릅니다.
"어 프레젠터 없는데 그냥 액티비티에 대충 짜야지 ㅎㅎ"

ii2001 replied:
제가 이해한게 맞다면 딱히 기능이 없더라도 구현자체는 해놓고 내부 클래스 내용이 비어있는게 아키텍처를 지키는 구조에서 더 좋은거라고 말씀 하신 것 같은데

그렇게 될 시에 오히려 내부에 사용하는 함수가 없어서 사람들이 불필요한 작업을 하게 되는게 아닐까요..?

laco-dev replied:
지금 논의하고 있는 것은, 아키텍처를 구현함에 따라 발생하는 보일러 플레이트를 얘기하고 있는 것이 아니에요.

아키텍처를 만든 목적 자체가, 프로젝트의 전반적인 설계를 구조화 하고 일반화 하기 위함인데 단순히 기능이 없으니 아키텍처를 안지킨다. 와 같은 잘못된 것으로 이해하실까봐 남긴 것입니다.

만약 반대로 생각해서 하나의 기능이 있어서 프레젠터를 만들었다고 가정할게요.
그런데 그 기능이 개발하다보니 필요 없어서 지웠어요.
그러면 다시 AppCompatActivity 구현하고 나머지 인터페이스는 모두 제거할건가요?

laco-dev commented on app/src/main/java/woowacourse/movie/seat/TheaterSeatActivity.kt:
```diff
@@ -3,35 +3,39 @@ package woowacourse.movie.seat
     private lateinit var presenter: TheaterSeatPresenter
-    private val totalPrice: TextView by lazy { findViewById(R.id.total_price) }
+    private val binding: ActivityTheaterSeatBinding by lazy {
+        ActivityTheaterSeatBinding.inflate(
+            layoutInflater,
+        )
```
`ActivityTheaterSeatBinding.setContentView(...)` 를 사용할 수 있을거에요

ii2001 replied:
dataBinding Util로 만들어서 해결했습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/seat/TheaterSeatActivity.kt:
```diff
@@ -74,7 +78,7 @@ class TheaterSeatActivity : AppCompatActivity(), TheaterSeatContract.View {
 
     override fun updateTotalPrice(price: Int) {
-        totalPrice.text = getString(R.string.total_price, price)
+        binding.invalidateAll()
```
price는 실제로 사용되지 않는데 invalidateAll() 동작이 왜 필요한가요?

ii2001 replied:
매개 변수를 지웠습니다! 예전 코드를 리펙토링하면서 까먹었네요 감사합니다!

laco-dev commented on app/src/main/java/woowacourse/movie/seat/TheaterSeatContract.kt:
```diff
@@ -42,6 +43,10 @@ interface TheaterSeatContract {
         )
 
-        fun getSelectedSeatNumbers(): String
+        fun getSelectedSeatNumbers(): Array<String>
+
+        fun getTotalPrice(): Int
```
마찬가지로 프레젠터나 뷰가 값을 반환하지 않도록 만들어주세요

laco-dev commented on app/src/main/res/layout/activity_movie_detail.xml:
```diff
@@ -0,0 +1,145 @@
+            type="woowacourse.movie.movieDetail.MovieDetailContract.Presenter" />
+    </data>
+
+    <androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+        xmlns:app="http://schemas.android.com/apk/res-auto"
+        xmlns:tools="http://schemas.android.com/tools"
```
네임스페이스는 모든 태그의 최상단에 작성하는것이 일반적입니다.

laco-dev commented on app/src/main/res/layout/activity_movie_detail.xml:
```diff
@@ -0,0 +1,145 @@
+                    android:layout_height="wrap_content"
+                    android:layout_gravity="center_horizontal"
+                    android:layout_marginTop="16dp"
+                    android:text="@{presenter.getMovie().title.toString()}"
```
값을 출력하기 위해서라도 프레젠터를 꺼내쓰는 것은 맞지 않아요.
MVP와 데이터바인딩의 한계지만, View가 바인딩 값을 업데이트 하도록 만들어주세요.


```
<variable movie: Movie />
text="@{movie.xxx}"
```

ii2001 replied:
수업에서 들은 내용으로는

`presenter를 참조를 안하고 View내부에서 하게되면 필요한 데이터를 View가 알게`

되는거니깐 View의 역할을 넘어간 부분이라고 생각
페로로 말씀대로 presenter만 데이터를 가지고 있는게 좋은 것 같습니다!
하는데 아직 MVP패턴을 이해를 잘 못한건지 잘 모르겠습니다!

laco-dev replied:
조금 더 경험과 학습이 필요한 부분이라 간단하게만 남길게요.

MVVM 아키텍처가 적용되어야 비로소 데이터바인딩이 본래의 역할을 한다고 볼 수 있어요.
지금은, 너무 깊게 이해하기 보다는 MVP의 원칙에 맞게 반환하지 않도록 만들어 볼 수 있겠습니다.

laco-dev reviewed (changes requested):
데이터바인딩으로 인해 MVP 구조가 잘못되었는지 아니면 잘못 이해하고 있는 부분인지는 정확히 모르겠어요.

다만 MVP의 구조에 원칙에 맞도록 잘 설계하는 것이 중요해 보입니다.
전반적으로 안드로이드 구현을 어느정도 하실 줄 아는 것으로 보여서 아키텍쳐를 비롯해 몇가지 고민해볼만한 내용으로 남겼습니다.

Commit:
:truck: 테스트 코드 패키지 프로덕트와 동일하게 맞춤

Commit:
:memo: README.md 할 일 전부 체크 완료

Commit:
:recycle: 프레젠터나 뷰가 값을 반환하지 않음

Commit:
:recycle: 네임스페이스는 모든 태그의 최상단에 작성

Commit:
:recycle: MovieDetailContract.kt 반환값 모두 삭제

Commit:
:recycle: onViewCreated() 다이어트

Commit:
:recycle: ktFormating

Commit:
:white_check_box: :recycle: 화면 이동 테스트 IntegrationTest.kt로 이동

Commit:
:sparkles: Intent오류 발생시 오류 Activity로 이동

Commit:
:sparkles: Activity, Fragment 바인딩기능 추가

Commit:
:recycle: Activity, Fragment 바인딩기능 적용

Commit:
:recycle: Cinema 프레젠터 내부로 이동

Commit:
:recycle: onClicked renameing

Commit:
:sparkles: Fragment redirectToErrorActivity

Commit:
:sparkles: TheatersBottomSheet MVP 구조 적용

Commit:
:recycle: MovieListPresenter.kt
loadMovie동작 프레젠터 내에서 init으로 적용

Commit:
:truck: showBottomSheet->navigateToCinemaView

Commit:
:truck: generateDateRange->loadDateRange

ii2001 commented:
현재 페어프로그래밍을 했던 크루들끼리 모두 같이 하고 있습니다! 
그래서 우테코에서 처음 안드로이드 프레임 워크를 배운 기간대비 많이 성장해서 여러가지 스킬들을 사용할 수 있었던 것 같아요!
다음 리뷰도 잘 부탁드리겠습니다!!! 🥰

Commit:
:recycle: ktFormat

Commit:
:bug: init으로 옯긴 후 private val 위치 변경

laco-dev commented on app/src/androidTest/java/woowacourse/movie/IntegrationTest.kt:
```diff
@@ -0,0 +1,77 @@
+
+class IntegrationTest {
+    @Test
+    fun `좌석선택버튼클릭_좌석선택표시`() {
```
> 좌석선택버튼클릭_좌석선택표시

만약 이 테스트가 실패했다면 표시가 되어야 하는지 되지 않아야 하는지 모르지 않을까요?

laco-dev commented on app/src/main/java/woowacourse/movie/ErrorActivity.kt:
```diff
@@ -0,0 +1,18 @@
+package woowacourse.movie
```
오류 화면도 패키지를 나눠보면 좋을 것 같아요

laco-dev commented on app/src/main/java/woowacourse/movie/common/BindingActivity.kt:
```diff
@@ -0,0 +1,18 @@
+package woowacourse.movie.common
```
common 보다는 base가 어울리지 않을까요?

laco-dev commented on app/src/main/java/woowacourse/movie/common/ui/ContextExtensions.kt:
```diff
@@ -0,0 +1,25 @@
+import androidx.fragment.app.Fragment
+import woowacourse.movie.ErrorActivity
+
+fun Activity.redirectToErrorActivity() {
+    val intent =
+        Intent(this, ErrorActivity::class.java).apply {
+            putExtra("ERROR_MESSAGE", "필요한 데이터를 불러오는 데 실패했습니다.")
+        }
+    startActivity(intent)
+    finish()
+}
+
+fun Fragment.redirectToErrorActivity() =
+    with(requireActivity()) {
+        val intent =
+            Intent(this, ErrorActivity::class.java).apply {
+                putExtra("ERROR_MESSAGE", "필요한 데이터를 불러오는 데 실패했습니다.")
+            }
+        startActivity(intent)
+        finish()
+    }
```
이 코드들도 마찬가지로 ErrorActivity, ErrorFragment가 스스로 하는게 좋아 보이는데 확장 함수로 빠진 이점이 무엇이 있나요?

ii2001 replied:
ErrorActivity 의 companion object 안에 넣어두겠습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailContract.kt:
```diff
@@ -1,34 +1,27 @@
+        fun onTicketPlus()
 
-        fun getMovie(): MovieInfo
+        fun onTicketMinus()
```
onXXX() 형태도 콜백을 의미하는 뜻이에요.
아래는 기능 중심으로 작성했을 때의 예시입니다.

```kotlin
increaseTicketCount()
decreaseTicketCount()
```

laco-dev commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailContract.kt:
```diff
@@ -1,34 +1,27 @@
+        fun onTicketMinus()
 
-        fun load(movie: MovieInfo)
+        fun updateTimes()
```
updateTimes() 이라는 인터페이스를 호출한다면 어떤 동작을 하기를 기대할까요?

시간을... 업데이트? 왜? 무엇으로?

ii2001 replied:
updateMovieRunTime으로 변경했습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailPresenter.kt:
```diff
@@ -1,47 +1,41 @@
+    var ticketNum = 1
 
-    private lateinit var movie: MovieInfo
+    val movie: MovieInfo = cinema.theater.movie
```
Presenter가 외부로 상태를 노출시키지 않도록 요청드렸는데, 오히려 더 늘어난 이유는 무엇인가요?

laco-dev replied:
이유를 물어봤는데..요 🤔

laco-dev commented on app/src/main/java/woowacourse/movie/movieList/MovieListPresenter.kt:
```diff
@@ -12,6 +12,13 @@ import woowacourse.movie.model.theater.Theater
 
 class MovieListPresenter(private val view: MovieListView) {
+
+    private val theaters: List<Theater> = generateTheaters(GENERATE_DUMMY_DATA_NUM)
```
프레젠터를 더욱 테스트 하기 좋은 구조로 만든다면, 이 값을 생성자로 받도록 해 볼 수도 있겠네요.

laco-dev commented on app/src/main/java/woowacourse/movie/movieList/cinemaListDialog/TheatersBottomSheetContract.kt:
```diff
@@ -0,0 +1,18 @@
+import woowacourse.movie.model.Cinema
+import woowacourse.movie.model.theater.Theater
+
+interface TheatersBottomSheetContract {
```
TheatersBottomSheetContract

따로 바텀시트라고 해서 BottomSheetContract라고 짓지는 않는 편이에요.
Contract 라는 인터페이스 입장에서는 구현하는 View가 Activity이던 Fragment이던 BottomSheet던지 관심 없으니까요

laco-dev commented on app/src/main/java/woowacourse/movie/movieList/cinemaListDialog/TheatersBottomSheetFragment.kt:
```diff
@@ -7,16 +7,31 @@ import android.view.View
+    private val presenter: TheatersBottomSheetContract.Presenter by lazy {
+        TheatersBottomSheetPresenter(this)
+    }
+    private lateinit var theater: Theater
```
프레젠터를 초기화 할 때 넣어주면 되는데, 극장 필드를 프레젠터가 아닌 뷰가 가지고 있어야만 하는 이유가 있나요?

laco-dev commented on app/src/main/java/woowacourse/movie/movieList/cinemaListDialog/TheatersBottomSheetFragment.kt:
```diff
@@ -32,56 +47,11 @@ class TheatersBottomSheetFragment : BottomSheetDialogFragment() {
-                ),
-            ),
-        )
+        adapter = CinemaAdapter { presenter.selectCinema(it) }
```
코틀린에서는 파라미터로 사용하는 고차 함수의 괄호를 생략할 수 있는 장점이 있는데 실제로는 이 함수가 무엇인지 오해를 부르는 경우가 종종 있습니다.

사용하는 입장에서는 이 고차함수가 그래서 무엇인데? 라는 의문이죠
개인적인 의견이지만, 저는 단순 함수형 인터페이스가 아닌 이상 항상 파라미터 이름을 사용합니다.
```kotlin
CinemaAdapter(onItemClick = { ... })
```

laco-dev commented on app/src/main/java/woowacourse/movie/movieList/cinemaListDialog/TheatersBottomSheetFragment.kt:
```diff
@@ -92,7 +62,12 @@ class TheatersBottomSheetFragment : BottomSheetDialogFragment() {
     }
 
+    override fun showCinemas(cinemas: List<Cinema>) {
+        binding.rvCinema.adapter = adapter
```
어댑터를 매번 다시 설정할 필요가 있나요?

laco-dev commented on app/src/main/java/woowacourse/movie/seat/TheaterSeatActivity.kt:
```diff
@@ -92,13 +88,13 @@ class TheaterSeatActivity : AppCompatActivity(), TheaterSeatContract.View {
                     binding.presenter = presenter
                 }
-        }
+        } ?: redirectToErrorActivity()
```
우리가 기대하는 것은 프레젠터가 초기화 하는데 실패하면 더이상 흐름이 진행되지 않고 화면이 종료되기를 생각할거에요.

그런데 이 함수 호출하는 다음 코드는 정상적으로 흘러가게 됩니다.
실제로 setupSeats() 코드가 이어서 호출되는 것을 확인할 수 있습니다.

물론 View 초기화 하는 것 자체는 문제가 되지 않을 수 있지만 초기화 하는데 실패했는데 다음 코드에 이어서 latetinit을 접근한다거나 하면 문제가 생길 수 있어요

laco-dev reviewed (changes requested):
아직 MVVM 패턴을 배우지 않았기 때문에, 데이터 바인딩을 온전히 이해하기는 다소 어려운 부분이 있습니다.

그렇기 때문에 MVP 패턴을 구현하는 만큼 데이터바인딩 이전에 MVP 패턴의 원칙을 지키도록 하는게 중요합니다.
프레젠터가 상태를 노출시키지 않도록 해서 MVP의 올바른 흐름대로 수정 해주세요.

Commit:
:recycle: 어댑터 구조 변경

Commit:
:recycle: 인텐트 전달 방식 변경

murjune committed:
:recycle: SeatPresenterTest capture 도입

murjune committed:
:recycle: fragment Container 이름 변경

murjune committed:
:recycle: timeDate formatting 개선

murjune committed:
:recycle: Time/Date Adapter 개선

murjune committed:
:recycle: ktlintCheck

murjune committed:
:truck: BindingActivity base package로 이동

murjune committed:
:recycle: ErrorActivity 시작 로직으로 변경

murjune committed:
:bug: Time/Date items MutableList로 변경

murjune committed:
:recycle: MovieDetailContract Presenter 네이밍 변경

- onTicketPlus -> increaseTicketCount
- onTicketMinus ->  decreaseTicketCount
- updateTimes -> updateRunMovieTimes
- loadDateRange -> loadRunMovieDateRange

Commit:
:truck: 바텀시트 파일 명확하게 변경

Commit:
:recycle: notifyDataSetChanged 추가

Commit:
:recycle: 불필요한 검증 제거

Commit:
:recycle: 데이터 넘겨주는 포맷 생성

Commit:
:recycle: slot -> MutableList

Commit:
:recycle: Presenter 생성자로 theaters 이동

Commit:
:truck: error 패키지 설정

Commit:
:recycle: 광고 처리 로직 변경

Commit:
:recycle: presenter 생성할때 theater바로 넘겨주기

Commit:
:recycle: onTheaterClicked 표시해주기!

Commit:
:recycle: adapter초기화 1회만

Commit:
:recycle: Presenter 비정상 종료시 finish되게 함

Commit:
:recycle: dataBinding 적용

Commit:
:recycle: dataBinding presenter 종속이 아닌 view만 알고있게 적용

Commit:
:recycle: ktLint Formating

Commit:
:recycle: Test File 함수변경에 따른 변경

Commit:
:bug: Test File 색 변경 잘못되어있는 부분 고침

murjune committed:
:recycle: TicketCountListener 로 네이밍 변경 및 패키지 분리

murjune committed:
:recycle: navigateToPurchaseConfirmation mvp 구조로 리팩토링

murjune committed:
:sparkles: MovieListMapper

murjune committed:
:sparkles: MovieDisplay UiModel 적용

murjune committed:
:recycle: SpinnerAdapter 객체 분리

- SpinnerDateAdapter
- SpinnerTimeAdapter

laco-dev commented on app/src/main/java/woowacourse/movie/error/ErrorActivity.kt:
```diff
@@ -15,4 +18,11 @@ class ErrorActivity : AppCompatActivity() {
     }
+
+    companion object {
+        @JvmStatic
```
JvmStatic 어노테이션은 왜 추가되었나요?

ii2001 replied:
실수로 지우지 않은 것 같습니다 삭제했습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/model/ui/MovieDisplay.kt:
```diff
@@ -0,0 +1,3 @@
+package woowacourse.movie.model.ui
+
+sealed class MovieDisplay(val viewType: Int)
```
이 타입은 목록 화면에서 사용하기 위한 개념으로 작성된 모델이 아닌가요?
실제로도 ViewType의 목적이 어댑터에 사용하기 위한 것이라면, 이것은 UI 모델이라고 단정할 수 있을까요?

UI 모델은 도메인과 1:1 대응되는 관계의 성격이라면, 아이템에 작성하기 위한 타입은 해당 화면에서 따로 작성을 하는 방법을 사용합니다.

```kotlin
sealed class MovieItem {
   data class Movie(val movie: Movie) : MovieItem()

   data class Ad(...): MovieItem()
}
```

laco-dev commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailActivity.kt:
```diff
@@ -1,97 +1,141 @@
+        presenter = MovieDetailPresenter(this, cinema)
+        initView()
+        initClickListener()
+        presenter.loadMovieInfo()
+        presenter.loadRunMovieDateRange()
```
프레젠터의 init 블록에서 이 두 함수를 호출하는 것과 어떤 차이가 있을까요?

ii2001 replied:
Activity자체가 생성될때 하는 것이 아닌 Create되는 생명주기때 작동하기 때문에 작동시기가 다를 것 같습니다.
현재로서는 같은 기능을 할 것 같긴 합니다

laco-dev commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailActivity.kt:
```diff
@@ -1,97 +1,141 @@
+
+    override fun showReleaseDate(movieDate: MovieDate) {
+        binding.date = movieDate.date
+        binding.invalidateAll()
```
아직 반영되지 않은 Binding을 명시적으로 적용할 때에는 binding.executePendingBindings()을 사용할 수 있습니다.
invalidateAll은 모든 동작을 일괄적으로 처리하기 때문에 다소의 차이가 있어요

laco-dev commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailActivity.kt:
```diff
@@ -1,97 +1,141 @@
+        }
+    }
+
+    private fun timeDate() =
+        buildString {
+            append(binding.movieDateSpinner.selectedItem.toString())
+            append(DELIMITER)
+            append(binding.movieTimeSpinner.selectedItem.toString())
+        }
```
다음 화면에서 그릴 데이터를 왜 이전화면이 결정하나요?

ii2001 replied:
다음화면에 보내줄 정보들이라 전화면이 가지고 있는게 맞다고 생각했습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailActivity.kt:
```diff
@@ -1,97 +1,141 @@
+            context: Context,
+            cinema: Cinema,
+        ): Intent {
+            return Intent(context, MovieDetailActivity::class.java).apply {
+                putExtra(EXTRA_CINEMA, cinema)
+            }
```
putExtra는 Intent 자신을 반환하기 때문에 Intent.putExtra() 처럼 작성할 수 있습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailContract.kt:
```diff
@@ -1,27 +1,43 @@
+        fun navigateToPurchaseConfirmation(cinema: Cinema)
 
-        fun onTicketCountChanged()
+        fun onTicketCountChanged(ticketNum: Int)
```
onTicketCountChanged() 랑 showTicketCount()는 어떤 차이가 있나요?

ii2001 replied:
Oh.. show가 더 좋아보이네요 확실히 이해가 잘됩니다..! 그렇게 변경하였습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/movieDetail/MovieDetailPresenter.kt:
```diff
@@ -46,4 +47,16 @@ class MovieDetailPresenter(
+
+    override fun loadMovieInfo() {
+        val movie = cinema.theater.movie
+        view.showTitle(movie.title)
+        view.showReleaseDate(movie.releaseDate)
+        view.showSynopsis(movie.synopsis)
+        view.showRunningTime(movie.runningTime)
```
각각 개별적으로 호출될 일이 없다면 필요한 데이터를 객체로 묶어서 주고 받는 방법도 있지 않을까요?

ii2001 replied:
객체 자체를 묶어서 보내는 것 보다 값 하나씩 보내는게 xml자체에서 데이터의 타입이 객체가 아니라서 더 좋다고 판단했습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/movieList/MovieAdapter.kt:
```diff
@@ -1,86 +1,81 @@
+        MovieViewHolder(binding.root) {
+        fun bind(movie: MovieItemDisplay) {
+            binding.data = movie
+            binding.listener = MovieItemListener { onClick(absoluteAdapterPosition) }
```
리스너를 데이터바인딩 변수로 전달하는 것과 `binding.button.setOnClickListener { ... }` 로 작성하는 것은 어떤 차이가 있나요?

리스너를 실제로 만들 필요가 있는 것인가에 대한 의문입니다.

ii2001 replied:
리스너를 나중에 또 사용할일이있다면 편리하게 또 가져다 쓸 수 있어서 이렇게 만들어 보았습니다..!

laco-dev commented on app/src/main/java/woowacourse/movie/movieList/MovieListMapper.kt:
```diff
@@ -0,0 +1,20 @@
+import woowacourse.movie.model.ui.MovieDisplay
+import woowacourse.movie.model.ui.MovieItemDisplay
+
+fun List<Theater>.toMovieDisplays(): List<MovieDisplay> {
+    return map(Theater::toMovieItemDisplay)
+        .chunked(3)
+        .flatMap { it + AdItemDisplay() }
+}
```
3개씩 쪼개는 것은 누가 결정해야할 책임인가요?
이 조건은 비즈니스 로직이 아닌가요?

ii2001 replied:
어,, Mapper가 비즈니스 로직을 처리하는 공간 아닌가요..?

laco-dev commented on app/src/main/java/woowacourse/movie/movieList/MovieListView.kt:
```diff
@@ -1,10 +1,10 @@
     fun showToast(message: String)
 
-    fun updateAdapter(displayData: List<MovieDisplayData>)
+    fun updateAdapter(displayData: List<MovieDisplay>)
```
프레젠터가 뷰의 어댑터를 초기화 하는지 알 필요가 있나요? showDisplayDate() 라고 작성한 것과 어떤 차이가 있을지 궁금합니다.

laco-dev commented on app/src/main/java/woowacourse/movie/seat/TheaterSeatActivity.kt:
```diff
@@ -86,18 +86,27 @@ class TheaterSeatActivity : BindingActivity<ActivityTheaterSeatBinding>(R.layout
+    private fun initPresenter() {
+        val ticketNum = intent.getStringExtra(EXTRA_TICKET_NUM) ?: return ErrorActivity.start(this)
+        val cinema = IntentCompat.getSerializableExtra(intent, EXTRA_CINEMA, Cinema::class.java)
+        if (cinema == null) {
+            ErrorActivity.start(this)
+            return finish()
+        }
```
여기서 finish()를 하더라도 호출하던 onCreate()함수가 종료되지는 않습니다.
그렇기 때문에 명확한 논리적인 상태에 따라 종료여부를 결정해야 한다면, 함수가 의도치 않게 이어지지 않도록 주의해야 합니다.

가령 initPresenter()가 종료되더라도 initSeats()는 이어서 호출됩니다.

ii2001 replied:
어.. 그렇다면 어떤식으로 종료를 해야 하는 것인가요..?

laco-dev commented on app/src/main/java/woowacourse/movie/seat/TheaterSeatPresenter.kt:
```diff
@@ -31,16 +33,15 @@ class TheaterSeatPresenter(
-                selectedSeats.add(seatId)
-            }
+            if (selectedSeats.size >= ticketLimit) return
+            seat.chosen = true
```
선택된 시트를 따로 관리하는데 chosen을 도메인에 포함하는 이유가 있나요?

ii2001 replied:
중복 코드였네요..! 삭제했습니다.

laco-dev commented on app/src/main/res/layout/activity_movie_detail.xml:
```diff
@@ -3,8 +3,28 @@
         <variable
-            name="presenter"
-            type="woowacourse.movie.movieDetail.MovieDetailPresenter" />
+            name="title"
+            type="String" />
+
+        <variable
+            name="runningTime"
+            type="String" />
+
+        <variable
+            name="synopsis"
+            type="String" />
+
+        <variable
+            name="date"
+            type="java.time.LocalDate" />
+
+        <variable
+            name="ticketNum"
+            type="Integer" />
```
이 정보를 가지고 있는 객체가 있다면 그 객체를 변수로 할당하면 되지 않을까요?

ii2001 replied:
객체 자체를 xml에서 가지고 있는거면 presenter자체를 가지고 있는것과 차이점이 있나요??

laco-dev commented on app/src/main/res/layout/fragment_movie_list.xml:
```diff
@@ -8,6 +9,7 @@
 
         <androidx.recyclerview.widget.RecyclerView
             android:id="@+id/rv_movies"
+            app:layoutManager="androidx.recyclerview.widget.LinearLayoutManager"
```
xml에 대해서도 린트가 돌지 않더라도 코드 정렬 컨벤션이 있습니다.
리포매팅 신경 써주세요

laco-dev commented on app/src/main/res/layout/item_ad.xml:
```diff
@@ -1,15 +1,23 @@
+            type="woowacourse.movie.model.ui.AdItemDisplay" />
+    </data>
+    <androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+        xmlns:app="http://schemas.android.com/apk/res-auto"
```
네임스페이스는 xml의 최상단 루트에 작성합니다.
지금 파일의 경우 layout 태그에 작성되어야 합니다.

laco-dev commented on app/src/main/res/layout/item_movie_list.xml:
```diff
@@ -1,55 +1,69 @@
+
+        <ImageView
+            android:id="@+id/movie_thumbnail"
+            android:layout_width="100dp"
+            android:layout_height="150dp"
```
이미지가 ImageView의 크기보다 작거나 크면 어떻게 그려지나요?

ii2001 replied:
이 부분은 그림의 크기를 저렇게 늘이고 축소를 하고 싶어서 지정했는데 다른 방법으로 해야하는 것일까요?

laco-dev reviewed (approved):
피드백 반영 고생하셨습니다.

나름 이유가 궁금해서 남긴 질문들이 많은데, 단순히 이모지로 답변을 해주시면 어떤 의도를 가지고 작성했는지 저는 알 방법이 없어진답니다. 😢 

🚀 ❤️ 👍🏻 ❤️ 🚀 해주시면 됩니다.

도메인적인 설계 보다는 다음 미션에서도 이어서 안드로이드 컴포넌트를 집중적으로 경험하게 되니 사용해야할 리소스를 잘 고민 해주시면 됩니다.
